### ISSUE TITLE
Warning in release mode with no MPI

### ISSUE THREAD

**Issue by rezarastak** [dealii/dealii#9820]
Related to #9613. I am observing these warnings while compiling in release mode with no MPI with gcc 7.5.0.
```
$ ninja obj_base_release 
[50/93] Building CXX object source/base/CMakeFiles/obj_base_release.dir/mpi_noncontiguous_partitioner.cc.o
In file included from ../include/deal.II/lac/vector.h:24:0,
                 from ../include/deal.II/lac/vector_memory.h:26,
                 from ../include/deal.II/lac/lapack_full_matrix.h:27,
                 from ../include/deal.II/base/tensor.h:29,
                 from ../include/deal.II/base/symmetric_tensor.h:25,
                 from ../include/deal.II/base/array_view.h:23,
                 from ../source/base/mpi_noncontiguous_partitioner.cc:16:
../include/deal.II/base/index_set.h: In member function ‘void dealii::Utilities::MPI::NoncontiguousPartitioner<Number>::reinit(const std::vector<unsigned int>&, const std::vector<unsigned int>&, const MPI_Comm&) [with Number = double]’:
../include/deal.II/base/index_set.h:1477:48: warning: ‘return_value’ may be used uninitialized in this function [-Wmaybe-uninitialized]
   , largest_range(numbers::invalid_unsigned_int)
                                                ^
In file included from ../include/deal.II/base/mpi_consensus_algorithms.h:22:0,
                 from ../include/deal.II/base/mpi_compute_index_owner_internal.h:22,
                 from ../include/deal.II/base/mpi_noncontiguous_partitioner.h:22,
                 from ../source/base/mpi_noncontiguous_partitioner.cc:17:
../include/deal.II/base/mpi.templates.h:338:9: note: ‘return_value’ was declared here
       T return_value;
         ^~~~~~~~~~~~
In file included from ../include/deal.II/lac/vector.h:24:0,
                 from ../include/deal.II/lac/vector_memory.h:26,
                 from ../include/deal.II/lac/lapack_full_matrix.h:27,
                 from ../include/deal.II/base/tensor.h:29,
                 from ../include/deal.II/base/symmetric_tensor.h:25,
                 from ../include/deal.II/base/array_view.h:23,
                 from ../source/base/mpi_noncontiguous_partitioner.cc:16:
../include/deal.II/base/index_set.h: In member function ‘void dealii::Utilities::MPI::NoncontiguousPartitioner<Number>::reinit(const std::vector<unsigned int>&, const std::vector<unsigned int>&, const MPI_Comm&) [with Number = float]’:
../include/deal.II/base/index_set.h:1477:48: warning: ‘return_value’ may be used uninitialized in this function [-Wmaybe-uninitialized]
   , largest_range(numbers::invalid_unsigned_int)
                                                ^
In file included from ../include/deal.II/base/mpi_consensus_algorithms.h:22:0,
                 from ../include/deal.II/base/mpi_compute_index_owner_internal.h:22,
                 from ../include/deal.II/base/mpi_noncontiguous_partitioner.h:22,
                 from ../source/base/mpi_noncontiguous_partitioner.cc:17:
../include/deal.II/base/mpi.templates.h:338:9: note: ‘return_value’ was declared here
       T return_value;
         ^~~~~~~~~~~~
[51/93] Building CXX object source/base/CMakeFiles/obj_base_release.dir/partitioner.cc.o
../source/base/partitioner.cc: In member function ‘void dealii::Utilities::MPI::Partitioner::set_ghost_indices(const dealii::IndexSet&, const dealii::IndexSet&)’:
../source/base/partitioner.cc:177:65: warning: ‘return_value’ may be used uninitialized in this function [-Wmaybe-uninitialized]
         Utilities::MPI::max(n_ghost_indices_data, communicator) > 0;
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
[93/93] Building CXX object source/base/CMakeFiles/obj_base_release.dir/data_out_base.cc.o
```
I was not able to figure out how to fix it, but I tracked the source of one of these warnings to #9600 using `git bisect`.

**Comment by masterleinad** [comment]
I can reproduce.

**Comment by peterrum** [comment]
@rezarastak Could you try out the following diff:
```diff
diff --git a/include/deal.II/base/mpi.templates.h b/include/deal.II/base/mpi.templates.h
index 1088896..e38eb26 100644
--- a/include/deal.II/base/mpi.templates.h
+++ b/include/deal.II/base/mpi.templates.h
@@ -230,7 +230,7 @@ namespace Utilities
     T
     sum(const T &t, const MPI_Comm &mpi_communicator)
     {
-      T return_value;
+      T return_value = t;
       internal::all_reduce(MPI_SUM,
                            ArrayView<const T>(&t, 1),
                            mpi_communicator,
@@ -335,7 +335,7 @@ namespace Utilities
     T
     max(const T &t, const MPI_Comm &mpi_communicator)
     {
-      T return_value;
+      T return_value = t;
       internal::all_reduce(MPI_MAX,
                            ArrayView<const T>(&t, 1),
                            mpi_communicator,
@@ -377,7 +377,7 @@ namespace Utilities
     T
     min(const T &t, const MPI_Comm &mpi_communicator)
     {
-      T return_value;
+      T return_value = t;
       internal::all_reduce(MPI_MIN,
                            ArrayView<const T>(&t, 1),
                            mpi_communicator,
```

Thanks!

**Comment by rezarastak** [comment]
@peterrum Your patch fixes the issue on my machine. No warnings.

Do you think there might be a performance penalty for initializing these variables?

**Comment by peterrum** [comment]
Nope. Let me create a PR.

**Comment by peterrum** [comment]
I am just wondering why #9600 caused the problem. It is not even related (only uses `Utilties::MPI::max()`).

**Comment by rezarastak** [comment]
Here I reported multiple warnings. only one of them
```
[51/93] Building CXX object source/base/CMakeFiles/obj_base_release.dir/partitioner.cc.o
../source/base/partitioner.cc: In member function ‘void dealii::Utilities::MPI::Partitioner::set_ghost_indices(const dealii::IndexSet&, const dealii::IndexSet&)’:
../source/base/partitioner.cc:177:65: warning: ‘return_value’ may be used uninitialized in this function [-Wmaybe-uninitialized]
         Utilities::MPI::max(n_ghost_indices_data, communicator) > 0;
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
```
is from #9600.

**Comment by kronbichler** [comment]
I guess `git bisect` is not perfect; we were moving things around in those files around that time.

**Comment by rezarastak** [comment]
This issue was solved by #9825. So I am closing it. Thank you @peterrum.

It is useful to have a dedicated tester for release mode to check for these warning (#9613).
